dev/src/mcal/Can.cpp

```diff
@@ -0,0 +1,237 @@
+
+    //  Set the bit timing register
+    CAN->BTR = (uint32_t)((uint32_t)conf.mode << 30) | \
+               ((uint32_t)conf.sjw << 24) | \
+               ((uint32_t)conf.bs1 << 16) | \
+               ((uint32_t)conf.bs2 << 20) | \
+               ((uint32_t)conf.buadRate - 1);
+
+    //  Request leave initialisation
+    SetOperatingMode(conf, kNormal);
+}
+void Can::FilterInit(const FilterConfig& conf) {
+    CHECK_FILTER_CONFIG();
+
+    CAN->FMR.FINIT = 1;  //  Initialisation mode for the filter
+    CAN->FA1R.registerVal &= ~(conf.bank);  //  Filter Deactivation
+
+    if (conf.scale == k16bit) {
+        CAN->FS1R.registerVal &= ~(conf.bank);
+        /* First 16-bit identifier and First 16-bit mask
+	       Or First 16-bit identifier and Second 16-bit identifier */
+        CAN->FilterRegister[conf.bank].FR1 = ((conf.maskIdLow & 0x0000FFFF)  << 16 | 
+                                              (conf.idLow & 0x0000FFFF));
+        /* Second 16-bit identifier and Second 16-bit mask
+	       Or Third 16-bit identifier and Fourth 16-bit identifier */                                
+        CAN->FilterRegister[conf.bank].FR2 = ((conf.maskIdHigh & 0x0000FFFF) << 16 | 
+                                              (conf.idHigh & 0x0000FFFF));
+    } else {
+        CAN->FS1R.registerVal |= (conf.bank);
+        //  32-bit identifier or First 32-bit identifier
+        CAN->FilterRegister[conf.bank].FR1 = ((conf.idHigh & 0x0000FFFF)     << 16 | 
+                                              (conf.idLow & 0x0000FFFF));
+        //  32-bit mask or Second 32-bit identifier
+        CAN->FilterRegister[conf.bank].FR2 = ((conf.maskIdHigh & 0x0000FFFF) << 16 | 
+                                              (conf.maskIdLow & 0x0000FFFF));
+    }
+    switch (conf.mode) {
+        case kMask: CAN->FM1R.registerVal &= ~(conf.bank); break;
+        case kList: CAN->FM1R.registerVal |=  (conf.bank); break;
+    }
+    switch (conf.fifoAssign) {
+        case kFIFO0: CAN->FFA1R.registerVal &= ~(conf.bank); break;
+        case kFIFO1: CAN->FFA1R.registerVal |=  (conf.bank); break;
+    }
+    if (conf.activation == kEnable) {
+        CAN->FA1R.registerVal |= (conf.bank); 
+    }
+    //  Leave the initialisation mode for the filter
+    CAN->FMR.FINIT = 0;
+}
+void Can::Transmit(CanTxMsg message) {
+    CHECK_MESSAGE_CONFIG();
+    uint32_t txMailbox = 0;
+    if ((CAN->TSR.TME0 != 0) || 
+        (CAN->TSR.TME1 != 0) ||
+        (CAN->TSR.TME2 != 0)) {
+            txMailbox = CAN->TSR.CODE;
+            if (txMailbox <= 2) {
+                if (message.ide == kStd) {
+                    CAN->TxMailBox[txMailbox].TIR.STID = message.stdId;
+                    CAN->TxMailBox[txMailbox].TIR.RTR = message.rtr;
+                } else {
+                    CAN->TxMailBox[txMailbox].TIR.STID = message.extId;
+                    CAN->TxMailBox[txMailbox].TIR.RTR = message.rtr;
+                    CAN->TxMailBox[txMailbox].TIR.IDE = 1;
```

Comment:
```
CAN->TxMailBox[txMailbox].TIR.STID = message.stdId;
CAN->TxMailBox[txMailbox].TIR.RTR = message.rtr;
if (message.ide != kStd)   CAN->TxMailBox[txMailbox].TIR.IDE = 1;
```
to avoid a lot of checking